### PR TITLE
First version of the URL Redirects (#9121)

### PR DESCRIPTION
**Pull request by mamazu**
| Q               | A
| --------------- | -----
| Branch?         | 1.0, 1.1 or master 
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | partially #9121 
| License         | MIT

As mentioned in the ticket above Sylius has no way of configuring redirects with its GUI. So I added an Entity and a grid to configure the Entity.

## Implementation
I have added an event listener to the kernel that checks if there is a redirect defined and if so, I return a redirect response to the new page.

### CODE REVIEW

File: app/migrations/Version20180130155207.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+{
+    /**
+     * @param Schema $schema
+     */
+    public function up(Schema $schema)
+    {
+        // this up() migration is auto-generated, please modify it to your needs
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() !== 'mysql', 'Migration can only be executed safely on \'mysql\'.');
+
+        $this->addSql('CREATE TABLE sylius_url_redirects (id INT AUTO_INCREMENT NOT NULL, oldRoute VARCHAR(255) NOT NULL, newRoute VARCHAR(255) NOT NULL, enabled TINYINT(1) NOT NULL, UNIQUE INDEX one_enabled_route (oldRoute, enabled), PRIMARY KEY(id)) DEFAULT CHARACTER SET utf8 COLLATE utf8_unicode_ci ENGINE = InnoDB');
```
Comment (stefandoorn) [inline app/migrations/Version20180130155207.php]: Index prevents multiple `oldRoute' entries with `enabled = 0`.

Comment (mamazu) [reply app/migrations/Version20180130155207.php]: Yes, that is indeed a problem. I have replaced it with a validator.

**Comment by Mozan** [comment]
ping

File: src/Sylius/Component/Core/Model/URLRedirect.php [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+     */
+    public function getType(): string
+    {
+        return $this->type;
+    }
+
+    /**
+     * @param string $type
+     */
+    public function setType(?string $type): void
```
Comment (lchrusciel) [inline src/Sylius/Component/Core/Model/URLRedirect.php]: Can't we simplify this logic?

File: src/Sylius/Bundle/AdminBundle/Resources/translations/messages.de.yml [hunk 3]
```diff
@@ -26,7 +26,8 @@ sylius:
                     shipping_methods: Versandarten
                     tax_categories: Steuerkategorien
                     tax_rates: Steuersätze
-                    zones: Gebiete
+                    url_rewrites: URL Umleitungen
```
[the remark this answers is not in the document]
Comment (mamazu) [reply src/Sylius/Bundle/AdminBundle/Resources/translations/messages.de.yml]: How would I do that?

File: src/Sylius/Bundle/AdminBundle/Resources/config/services.xml [hunk 4]
```diff
@@ -37,5 +37,7 @@
 
         <service id="sylius.http_client" class="GuzzleHttp\Client" public="false" />
         <service id="sylius.http_message_factory" class="Http\Message\MessageFactory\GuzzleMessageFactory" public="false" />
+
```
[the remark this answers is not in the document]
Comment (mamazu) [reply src/Sylius/Bundle/AdminBundle/Resources/config/services.xml]: I removed one line but it doesn't show as outdated.

File: src/Sylius/Bundle/AdminBundle/Resources/translations/messages.de.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (Zales0123) [reply src/Sylius/Bundle/AdminBundle/Resources/translations/messages.de.yml]: In **Sylius** repository we only manage english translations, to fix/add translations in other languages you should go to the https://crowdin.com/project/sylius and make some work there :)

[the remark this answers is not in the document]
Comment (mamazu) [reply src/Sylius/Bundle/AdminBundle/Resources/translations/messages.de.yml]: But the translation keys down show up in the Sylius project.

[the remark this answers is not in the document]
Comment (lchrusciel) [reply src/Sylius/Bundle/AdminBundle/Resources/translations/messages.de.yml]: From the perspective of the PR, we need to remove these German translations. We need to add them to crowdin, once they will appear there

**Comment by lchrusciel** [comment]
Can you also run `bin/ecs check src/Sylius/Bundle src/Sylius/Component --fix` it should resolve most of coding standard problems. Thanks!

**Comment by mamazu** [comment]
@lchrusciel Thanks for the tip.

**Comment by mamazu** [comment]
Is this still relevant?

**Comment by mamazu** [comment]
Looks ready to merge for me.

**Comment by stefandoorn** [comment]
I see there is also a redirect plugin out there: https://github.com/Setono/SyliusRedirectPlugin.

**Comment by mamazu** [comment]
@pamil What do you think?

**Comment by pamil** [comment]
Sorry for the late answer. From the maintainer's perspective, I think it's the best to create a dedicated plugin for that which can be later merged into the core or incorporated as an official plugin. I think that currently, as Sylius, we don't have enough resources to support another feature in the core.

**Comment by mamazu** [comment]
Okay then I will close this issue and refer to the plugin: [SyliusRedirectPlugin](https://github.com/Setono/SyliusRedirectPlugin)

**Pull request closed without merging**